src/iop/hazeremoval.c

```diff
@@ -0,0 +1,973 @@
+  dt_bauhaus_slider_set(g->strength, p->strength);
+  dt_bauhaus_slider_set(g->distance, p->distance);
+
+  dt_pthread_mutex_lock(&g->lock);
+  g->distance_max = NAN;
+  g->A0[0] = NAN;
+  g->A0[1] = NAN;
+  g->A0[2] = NAN;
+  g->hash = 0;
+  dt_pthread_mutex_unlock(&g->lock);
+}
+
+void gui_init(dt_iop_module_t *self)
+{
+  self->gui_data = malloc(sizeof(dt_iop_hazeremoval_gui_data_t));
+  dt_iop_hazeremoval_gui_data_t *g = (dt_iop_hazeremoval_gui_data_t *)self->gui_data;
+  dt_iop_hazeremoval_params_t *p = (dt_iop_hazeremoval_params_t *)self->params;
+
+  dt_pthread_mutex_init(&g->lock, NULL);
+  g->distance_max = NAN;
+  g->A0[0] = NAN;
+  g->A0[1] = NAN;
+  g->A0[2] = NAN;
+  g->hash = 0;
+
+  self->widget = gtk_box_new(GTK_ORIENTATION_VERTICAL, DT_BAUHAUS_SPACE);
+
+  g->strength = dt_bauhaus_slider_new_with_range(self, -1, 1, 0.01, p->strength, 2);
+  dt_bauhaus_widget_set_label(g->strength, NULL, _("strength"));
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->strength), TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->strength), "value-changed", G_CALLBACK(strength_callback), self);
+
+  g->distance = dt_bauhaus_slider_new_with_range(self, 0, 1, 0.005, p->distance, 3);
+  dt_bauhaus_widget_set_label(g->distance, NULL, _("distance"));
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->distance), TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->distance), "value-changed", G_CALLBACK(distance_callback), self);
+}
+
+void gui_cleanup(dt_iop_module_t *self)
+{
+  dt_iop_hazeremoval_gui_data_t *g = (dt_iop_hazeremoval_gui_data_t *)self->gui_data;
+  dt_pthread_mutex_destroy(&g->lock);
+  free(self->gui_data);
+  self->gui_data = NULL;
+}
+
+//----------------------------------------------------------------------
+// module local functions and structures required by process function
+//----------------------------------------------------------------------
+
+typedef struct tile {
+  int left, right, lower, upper;
+} tile;
+
+typedef struct rgb_image {
+  rgb_pixel *data;
+  int width, height, stride;
+} rgb_image;
+
+static inline float * get_rgb_pixel(rgb_image img, size_t i)
+{
+  return ((float *)img.data) + i * img.stride;
+}
+
+typedef float gray_pixel;
+
+typedef struct gray_image {
+  gray_pixel *data;
+  int width, height;
+} gray_image;
+
+// allocate space for 1-component image of size width x height
+static inline gray_image new_gray_image(int width, int height)
+{
+  return (gray_image){ dt_alloc_align(64, sizeof(gray_pixel) * width * height), width, height };
+}
+
+// free space for 1-component image
+static inline void free_gray_image(gray_image *img_p)
+{
+  dt_free_align(img_p->data);
+  img_p->data = NULL;
+}
+
+// copy 1-component image img1 to img2
+static inline void copy_gray_image(gray_image img1, gray_image img2)
+{
+  memcpy(img2.data, img1.data, sizeof(gray_pixel) * img1.width * img1.height);
+}
+
+// minimum of two integers
+static inline int min_i(int a, int b)
```

Comment:
As far as darktable sources are concerned, they are defined in glib:
```
[ 79%] Building C object src/iop/CMakeFiles/exposure.dir/introspection_exposure.c.o
In file included from /home/lebedevri/darktable/build/src/iop/introspection_exposure.c:29:0:
/home/lebedevri/darktable/src/iop/exposure.c:52:0: error: "MAX" redefined [-Werror]
 #define MAX MIN
 
In file included from /usr/lib/x86_64-linux-gnu/glib-2.0/include/glibconfig.h:9:0,
                 from /usr/include/glib-2.0/glib/gtypes.h:32,
                 from /usr/include/glib-2.0/glib/galloca.h:32,
                 from /usr/include/glib-2.0/glib.h:30,
                 from /home/lebedevri/darktable/src/common/introspection.h:21,
                 from /home/lebedevri/darktable/src/iop/iop_api.h:28,
                 from <command-line>:0:
/usr/include/glib-2.0/glib/gmacros.h:288:0: note: this is the location of the previous definition
 #define MAX(a, b)  (((a) > (b)) ? (a) : (b))
 
In file included from /home/lebedevri/darktable/build/src/iop/introspection_exposure.c:29:0:
/home/lebedevri/darktable/src/iop/exposure.c:53:0: error: "MIN" redefined [-Werror]
 #define MIN MAX
 
In file included from /usr/lib/x86_64-linux-gnu/glib-2.0/include/glibconfig.h:9:0,
                 from /usr/include/glib-2.0/glib/gtypes.h:32,
                 from /usr/include/glib-2.0/glib/galloca.h:32,
                 from /usr/include/glib-2.0/glib.h:30,
                 from /home/lebedevri/darktable/src/common/introspection.h:21,
                 from /home/lebedevri/darktable/src/iop/iop_api.h:28,
                 from <command-line>:0:
/usr/include/glib-2.0/glib/gmacros.h:291:0: note: this is the location of the previous definition
 #define MIN(a, b)  (((a) < (b)) ? (a) : (b))
```

`/usr/include/glib-2.0/glib/gmacros.h` contains
```
#undef	MAX
#define MAX(a, b)  (((a) > (b)) ? (a) : (b))

#undef	MIN
#define MIN(a, b)  (((a) < (b)) ? (a) : (b))
```

So unless those don't suit your code, please just use them.